### DISCUSSION TITLE
proposal: Use squash merging instead

### DISCUSSION THREAD

**Discussion by Xuanwo** [databendlabs/databend#3221] [Ideas]
## Background

Github support three merge methods:

- merge commit
- rebase merging
- squash merging

For merge commit:

![image](https://user-images.githubusercontent.com/5351546/144540079-21e0485c-6211-43f5-9d06-8fedea659067.png)

We will have a separate merge commit along with all commits in PR.

For rebase merging:

We will share the same history.

For squash merging:

![image](https://user-images.githubusercontent.com/5351546/144540208-d5b05423-8393-4ecc-a025-99b64c4c27bc.png)

We will merge the whole PR into a single commit.

For now, databend uses merge commit which makes it hard to track changes.

## Propose

So I propose to use `squash merging` only for mering.

Benefits:

- Contributors don't need to struggle with force-pushing commits.
- Reviewers can track changes inside PR more easily (we can compare changes between commit if contributors don't need to force push commits).
- We can track history more easily. 

Drawbacks:

> Lose information about when specific changes were originally made and who authored the squashed commits. 

But we can track that information in PR

> If you continue working on the head branch of a pull request after squashing and merging, and then create a new pull request between the same branches, commits that you previously squashed and merged will be listed in the new pull request. 

This requires all contributors to start working from the `main` branch.

> Some Git commands that use the "SHA" or "hash" ID may be harder to use since the SHA ID for the original commits is lost. 

### Why not rebase merging?

It's highly possible for contributors to force push their commits before merging. To an active community like databend, it's not affordable.

**Comment by dragonly** [comment]
> Why not rebase merging?

Here's another supporting facts: I saw a lot of PR with commit comments like "update", "push", "u", etc. These stash-ish style of commit message is really meaning less for the whole view of master branch history.

We can make greater comment in PR and the squash merge commit message all in once.

**Reply by dragonly** [reply]
Here's another reference of practice https://github.com/ClickHouse/ClickHouse/issues/16789
